xonsh/tools.py

```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
         """A lazy value getter."""
         return self._cmds_cache.get(key, default)
 
+    def locate_binary(self, name):
+        """Locates an executable on the file system."""
+
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+
+            local_bins = [full_name for full_name in self.get_possible_names(name)
+                          if os.path.isfile(full_name)]
+            if local_bins:
+                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self.all_commands.keys()
+
+        if matches:
+            return self.all_commands[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
+    def lazy_locate_binary(self, name):
+        """Locates an executable in the cache."""
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self._cmds_cache.keys()
```

Comment:
Yeah, it is the building of the set of keys that takes up the time, coupled with the fact that dictionary lookup is one of the fastest things that you can do in Python  Here is some simple timeit:

```
scopatz@artemis ~ $ timeit "next((x for x in range(1000) if x == 1), None)"
1000000 loops, best of 3: 600 ns per loop
scopatz@artemis ~ $ d = {i: i for i in range(1000)}
scopatz@artemis ~ $ p = {1}
scopatz@artemis ~ $ timeit "p & d.keys()"
100000 loops, best of 3: 11.5 µs per loop
scopatz@artemis ~ $ 11.5e3 / 600
19.166666666666668
scopatz@artemis ~ $ timeit "next((x for x in d if x == 1), None)"
1000000 loops, best of 3: 392 ns per loop
scopatz@artemis ~ $ 11.5e3 / 392
29.336734693877553
```

So this is about ~20x - 30x slower with a len-1000 sized cache.  Note that my cache is about 3600.